atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java

```diff
@@ -17,47 +17,107 @@
 package com.palantir.atlasdb.keyvalue.cassandra;
 
 import java.net.InetSocketAddress;
+import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
 import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
 
-public class Blacklist {
+class Blacklist {
     private static final Logger log = LoggerFactory.getLogger(CassandraClientPool.class);
 
-    private Map<InetSocketAddress, Long> blacklistedHosts = Maps.newConcurrentMap();
+    private final CassandraKeyValueServiceConfig config;
+    private Map<InetSocketAddress, Long> blacklist;
 
-    public Map<InetSocketAddress, Long> getBlacklistedHosts() {
-        return blacklistedHosts;
+    Blacklist(CassandraKeyValueServiceConfig config) {
+        this.config = config;
+        this.blacklist = Maps.newConcurrentMap();
     }
 
-    // TODO (gsheasby): Why did we blacklist this host?
-    public void add(InetSocketAddress badHost) {
-        blacklistedHosts.put(badHost, System.currentTimeMillis());
-        log.warn("Blacklisted host '{}'", SafeArg.of("badHost", CassandraLogHelper.host(badHost)));
+    void checkAndUpdate(Map<InetSocketAddress, CassandraClientPoolingContainer> pools) {
+        // Check blacklist and re-integrate or continue to wait as necessary
+        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : blacklist.entrySet()) {
+            if (coolOffPeriodExpired(blacklistedEntry)) {
+                InetSocketAddress host = blacklistedEntry.getKey();
+                if (isHostHealthy(pools.get(host))) {
+                    blacklist.remove(host);
+                    log.info("Added host {} back into the pool after a waiting period and successful health check.",
+                            SafeArg.of("host", CassandraLogHelper.host(host)));
+                }
+            }
+        }
     }
 
-    public void addAll(Set<InetSocketAddress> hosts) {
+    private boolean coolOffPeriodExpired(Map.Entry<InetSocketAddress, Long> blacklistedEntry) {
+        long backoffTimeMillis = TimeUnit.SECONDS.toMillis(config.unresponsiveHostBackoffTimeSeconds());
```

Comment:
Calculate this only once in the constructor?

Reply:
`CassandraKeyValueServiceConfigManager` actually supports live-reloading the `CassandraKeyValueServiceConfig`, and I suspect this is used by the large internal product. Let's leave it here.

But I also think we're breaking this live-reloading, since we're just passing the Config object around, instead of the ConfigManager. Filed #2839.